### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by lukrzak**

### CODE REVIEW

File: communication/smartContractHandler.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		}
+		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
+			http.Error(w, "Invalid input", http.StatusBadRequest)
+			return
+		}
+		var output struct {
+			Output string `json:"output"`
+		}
+
+		utils.VerifySignature(input.Signature, fmt.Sprintf("%v", input.Payload), input.Signature)
```
Comment (tpalinski) [inline communication/smartContractHandler.go]: sprintf potrafi spaść z rowerka przy tym jsonie. lepiej zrobić json marshall + string z []byte

File: communication/smartContractHandler.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+		if input.View {
+			cmd := exec.Command("/bin/dopechain-vm", "-e", input.Payload.Entrypoint, "-a", input.Payload.Args, "-s", input.Payload.Sender, "--block-number", fmt.Sprintf("%d", len(blockchain.DopeChain)-1))
+			out, err := cmd.CombinedOutput()
+			if err != nil {
+				log.Warnf("error while running VM: %s", err)
+			}
+			log.Infof("VM output: %s", out)
```
Comment (tpalinski) [inline communication/smartContractHandler.go]: masz cały package vm który to robi

File: communication/transactionsHandler.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	}
+}
+
+func beginTransaction(sender string, amount int, receiver string) {
+	transToSend := blockchain.Transaction{Amount: amount, Receiver: receiver, Sender: sender}
+	if sender == receiver {
+		log.Info("cannot send $ to yourself")
+		return
+	}
+	blockchain.DopeTransactions.SaveTransaction(&transToSend)
```
Comment (tpalinski) [inline communication/transactionsHandler.go:64]: check na to czy masz tyle na koncie?

File: utils/crypto.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+)
+
+func VerifySignature(public string, message string, signature string) (bool, error) {
+	block, _ := pem.Decode([]byte(public))
```
Comment (tpalinski) [inline utils/crypto.go]: nie zadziała, samo base64 na bajty powinno być

File: utils/crypto.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+)
+
+func VerifySignature(public string, message string, signature string) (bool, error) {
```
Comment (tpalinski) [inline utils/crypto.go]: signature powinien byc []byte

**Review by tpalinski — CHANGES_REQUESTED** [review]

File: communication/transactionsHandler.go (hunk 3 above)
Comment (lukrzak) [reply communication/transactionsHandler.go:64]: Jest sprawdzane podczas wydobycia bloku

**Comment by tpalinski** [comment]
How does the app not respect the config? Please get rid of all the flag stuff in main, otherwise it will be undeployable

File: communication/smartContractHandler.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		}
+		sig, err := base64.StdEncoding.DecodeString(input.Signature)
+		if err != nil {
+			http.Error(w, "Invalid signature encoding", http.StatusBadRequest)
+			return
+		}
+		utils.VerifySignature(sndr, string(marshalledPayload), sig)
+
+		if input.View {
+			out, err := vm.RunContract(&vm.RunContractOpts{BinaryPath: "/bin/dopechain-vm", Entrypoint: input.Payload.Entrypoint, Args: input.Payload.Args, Sender: string(input.Payload.Sender), TransactionId: blockchain.DopeTransactions[len(blockchain.DopeTransactions)-1].Id})
```
Comment (tpalinski) [inline communication/smartContractHandler.go]: get rid of this hardcoded binary, we have everything set up in config.toml

File: communication/smartContractHandler.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		if err != nil {
+			http.Error(w, "Invalid sender encoding", http.StatusBadRequest)
+			return
+		}
+		sig, err := base64.StdEncoding.DecodeString(input.Signature)
+		if err != nil {
+			http.Error(w, "Invalid signature encoding", http.StatusBadRequest)
+			return
+		}
+		utils.VerifySignature(sndr, string(marshalledPayload), sig)
```
Comment (tpalinski) [inline communication/smartContractHandler.go]: Why do we need those as strings actually?

File: communication/smartContractHandler.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		if input.View {
+			out, err := vm.RunContract(&vm.RunContractOpts{BinaryPath: "/bin/dopechain-vm", Entrypoint: input.Payload.Entrypoint, Args: input.Payload.Args, Sender: string(input.Payload.Sender), TransactionId: blockchain.DopeTransactions[len(blockchain.DopeTransactions)-1].Id})
+			if err != nil {
+				log.Warnf("error while running VM: %s", err)
+			}
+			log.Infof("VM output: %s", out)
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		json.NewEncoder(w).Encode(output)
```
Comment (tpalinski) [inline communication/smartContractHandler.go]: We are not setting output property? So always returning empty output?

File: communication/communication.go [hunk 9]
```diff
@@ -32,11 +36,16 @@ func ConnectToNetwork(bootstrapAddr *string, ip *string, port *int, url string)
...
-	insertNamespace()
+	insertNamespaces()
 
 	go func() {
 		nodeAddress := fmt.Sprintf("%s:%d", *ip, *port)
 		http.HandleFunc(NODE_ENDPOINT, nodeHandler)
+		http.HandleFunc(ACCOUNTS_ENDPOINT, handleAccounts)
+		http.HandleFunc(ACCOUNTS_INFO_ENDPOINT, handleAccountsInfo)
+		http.HandleFunc(TRANSFER_ENDPOINT, handleTransfer)
+		http.HandleFunc(SMARTCONTRACT_ENDPOINT, handleSmartContract)
```
Comment (tpalinski) [inline communication/communication.go:47]: You can also define endpoints like `POST /api/account` and it will handle only this method

**Review by tpalinski — CHANGES_REQUESTED** [review]
Please use config.toml

**Pull request merged**